src/main/java/org/alliancegenome/curation_api/jobs/executors/GeneOntologyAnnotationExecutor.java

```diff
@@ -32,88 +35,91 @@
-
-		// curie, List<GO curie>
-		Map<String, List<String>> uiMap = new HashMap<>();
-		GZIPInputStream stream = new GZIPInputStream(new FileInputStream(bulkLoadFileHistory.getBulkLoadFile().getLocalFilePath()));
-		try (BufferedReader br = new BufferedReader(new InputStreamReader(stream))) {
-			Stream<String> lines = br.lines();
-
-			// Process each line
-			lines.filter(s -> !s.startsWith("!") && StringUtils.isNotEmpty(s)).forEach(s -> {
-				String[] token = s.split("\t");
-				String orgID = token[0];
-				String modID = token[1];
-				String goID = token[4];
-				if (abbr.equalsIgnoreCase(orgID) || abbr.equals("XB") && orgID.equalsIgnoreCase("Xenbase") || abbr.equals("HUMAN") && orgID.equals("RGD")) {
-					List<String> goIDs = uiMap.computeIfAbsent(modID, list -> new ArrayList<>());
-					goIDs.add(goID);
-				}
-			});
+		BulkFMSLoad fmsLoad = (BulkFMSLoad) bulkLoadFileHistory.getBulkLoad();
+		BackendBulkDataProvider dataProvider = BackendBulkDataProvider.valueOf(fmsLoad.getFmsDataSubType());
 
-		} catch (IOException e) {
-			e.printStackTrace();
+		CsvSchema csvSchema = CsvSchemaBuilder.gafSchema();
+		CsvMapper csvMapper = new CsvMapper();
+		MappingIterator<GeneOntologyAnnotationDTO> it = csvMapper.enable(CsvParser.Feature.INSERT_NULLS_FOR_MISSING_COLUMNS).readerFor(GeneOntologyAnnotationDTO.class).with(csvSchema).readValues(new GZIPInputStream(new FileInputStream(bulkLoadFileHistory.getBulkLoadFile().getLocalFilePath())));
+		List<GeneOntologyAnnotationDTO> gafData = it.readAll();
+
+		List<String> gafHeaderData = new ArrayList<>();
+		for (GeneOntologyAnnotationDTO gafLine : gafData) {
+			if (gafLine.getDb().startsWith("!")) {
+				gafHeaderData.add(gafLine.getDb());
+			} else {
+				break;
+			}
 		}
+		gafData.subList(0, gafHeaderData.size()).clear();
 
 		String name = bulkLoadFileHistory.getBulkLoad().getName();
 
-		Map<Long, GeneOntologyAnnotationDTO> gafMap = service.getGafMap(abbr);
-		List<Long> gafIdsBefore = new ArrayList<>(gafMap.keySet().stream().toList());
-		gafIdsBefore.removeIf(Objects::isNull);
+		List<Long> gafIdsBefore = geneOntologyAnnotationService.getAllGafIdsPerProvider(dataProvider.name());
+		List<Long> gafIdsLoaded = new ArrayList<>();
 
-		List<Long> geneGoIdsLoaded = new ArrayList<>();
-		ProcessDisplayHelper ph = new ProcessDisplayHelper();
-		ph.addDisplayHandler(loadProcessDisplayService);
-		Set<GeneOntologyAnnotationDTO> dtos = uiMap.entrySet()
-			.stream()
-			.map(entry -> entry.getValue().stream().map(goID -> {
-				GeneOntologyAnnotationDTO dto = new GeneOntologyAnnotationDTO();
-				String prefix = abbr;
-				if (abbr.equalsIgnoreCase("XB")) {
-					prefix = "Xenbase";
-				}
-				if (abbr.equalsIgnoreCase("HUMAN") || abbr.equalsIgnoreCase("MGI")) {
-					prefix = null;
+		Map<String, Set<String>> annotationMap = new HashedMap<>();
+
+		for (GeneOntologyAnnotationDTO annotation : gafData) {
+			if (annotation.getDb().equals(dataProvider.resourceDescriptor)) {
```

Comment:
@markquintontulloch this check should work for the Xenbase and then later we use the "prefix" for the id which is XB: for all three types.

Reply:
Right, but the old code used Xenbase: for everything

Reply:
yes the DB aka first field in the file is the dataProvider.resourceDescriptor which is the text "Xenbase".

Reply:
Sorry again, was confused by old code setting fields in the DTO object rather than the database.